wizard/src/utils/obs_generation.ts

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Column,
+  CsvData,
+  MappedThing,
+  Mapping,
+  MappingType,
+  MappingVal,
+  Observation,
+  RowObservations,
+} from "../types";
+
+// Helper maps used while generating observaitons.
+interface ObsGenMaps {
+  // Column Index -> mapping thing found in header for VALUE.
+  valCol2Hdr: Map<number, MappedThing>;
+  // Mapped thing -> column index (for non-VALUE)
+  thing2Col: Map<MappedThing, number>;
+  // Mapped thing -> constant value (for non-VALUE)
+  thing2Const: Map<MappedThing, string>;
+}
+
+/**
+ * Given a mapping and CSV Data computes the observations for every row
+ * corresponding to |rowsForDisplay|.
+ * ASSUMES: checkMappings() returns success on |mappings|
+ */
+export function generateRowObservations(
+  mappings: Mapping,
+  csvData: CsvData
+): RowObservations {
+  // Compute ObsGenMaps first.
+  const obsGenMaps: ObsGenMaps = {
+    valCol2Hdr: new Map<number, MappedThing>(),
+    thing2Col: new Map<MappedThing, number>(),
+    thing2Const: new Map<MappedThing, string>(),
+  };
+  for (const mthing of Array.from(mappings.keys())) {
+    const mval: MappingVal = mappings.get(mthing);
+    if (mval.type === MappingType.COLUMN_HEADER) {
+      for (const hdr of mval.headers) {
+        obsGenMaps.valCol2Hdr.set(hdr.columnIdx, mthing);
+      }
+    } else if (mval.type == MappingType.COLUMN) {
+      if (mthing == MappedThing.VALUE) {
+        obsGenMaps.valCol2Hdr.set(mval.column.columnIdx, mthing);
+      } else {
+        obsGenMaps.thing2Col.set(mthing, mval.column.columnIdx);
+      }
+    } else if (mval.type == MappingType.CONSTANT) {
+      obsGenMaps.thing2Const.set(mthing, mval.constant);
+    }
+  }
+
+  const rowObs: RowObservations = new Map();
+  for (const idx of Array.from(csvData.rowsForDisplay.keys())) {
+    const row = csvData.rowsForDisplay.get(idx);
+    const obs = generateObservationsInRow(
```

Comment:
nit: functions should be defined before they are used, so `generateObservationsInRow` should be defined before this current function

Reply:
I see. Moved it around.